packages/api/src/car.js

```diff
@@ -107,18 +109,23 @@ export async function carPost (request, env, ctx) {
 export async function handleCarUpload (request, env, ctx, car, uploadType = 'Car') {
   const { user, authToken } = request.auth
   const { headers } = request
-  let structure = uploadType === 'Upload' ? 'Complete' : 'Unknown'
+
+  // load the car into memory just the once here
+  const carBytes = new Uint8Array(await car.arrayBuffer())
 
   // Throws if CAR is invalid by our standards.
   // Returns either the sum of the block sizes in the CAR, or the cumulative size of the DAG for a dag-pb root.
-  const { size: dagSize, rootCid } = await carStat(car)
-  const sourceCid = rootCid.toString()
+  const { size: dagSize, rootCid, structure } = await carStat(carBytes)
 
-  if (structure === 'Unknown' && rootCid.code === raw.code) {
-    structure = 'Complete'
-  }
+  const sourceCid = rootCid.toString()
+  const carCid = CID.createV1(0x202, await sha256.digest(carBytes))
+  const s3Key = `raw/${sourceCid}/${user._id}/${toString(carCid.multihash.bytes, 'base32')}.car`
+  const r2Key = `${carCid.toString()}/${carCid.toString()}.car`
```

Comment:
`toString` unnecessary and with it makes slightly more readable:

```suggestion
  const r2Key = `${carCid}/${carCid}.car`
```